### PR TITLE
Fetch gitlab state change events from API

### PR DESCRIPTION
**Pull request by eseliger**
Closes https://github.com/sourcegraph/sourcegraph/issues/15899

Also fixes pagination over more than 20 events and migrates the events away from the webhooks package, so these are now sanitized from payloads.

### CODE REVIEW

File: enterprise/internal/campaigns/webhooks_gitlab.go [hunk 1]
```diff
@@ -171,6 +171,9 @@ func (h *GitLabWebhook) handleEvent(ctx context.Context, extSvc *repos.ExternalS
 	// webhook.
 	case *webhooks.MergeRequestApprovedEvent,
 		*webhooks.MergeRequestUnapprovedEvent,
+		*webhooks.MergeRequestCloseEvent,
+		*webhooks.MergeRequestMergeEvent,
+		*webhooks.MergeRequestReopenEvent,
 		*webhooks.MergeRequestUpdateEvent:
```
Comment (eseliger) [inline enterprise/internal/campaigns/webhooks_gitlab.go:177]: This would mean that all reload now. It was just easier for now

File: enterprise/internal/campaigns/changeset_history.go [hunk 2]
```diff
@@ -105,6 +105,8 @@ func computeHistory(ch *campaigns.Changeset, ce ChangesetEvents) (changesetHisto
 	}
 	pushStates(openedAt)
 
+	var isDraft = currentExtState == campaigns.ChangesetExternalStateDraft
```
Comment (eseliger) [inline enterprise/internal/campaigns/changeset_history.go:108]: This is a fix I found along the way, should probably be moved to a separate PR, but it tripped me up a few times while testing

File: cmd/repo-updater/repos/gitlab.go [hunk 3]
```diff
@@ -354,6 +354,11 @@ func (s *GitLabSource) CreateChangeset(ctx context.Context, c *Changeset) (bool,
 		}
 	}
 
+	// These additional API calls can go away once we can use the GraphQL API.
+	if err := s.decorateMergeRequestData(ctx, project, mr); err != nil {
+		return exists, errors.Wrapf(err, "retrieving additional data for merge request %d", mr.IID)
+	}
```
Comment (eseliger) [inline cmd/repo-updater/repos/gitlab.go:360]: I think this should happen on every modifying call, like we have it for GitHub. Otherwise, we won't have state events right after a reopen or close operation for example.

File: cmd/repo-updater/repos/gitlab.go [hunk 4]
```diff
@@ -536,6 +542,40 @@ func readSystemNotes(it func() ([]*gitlab.Note, error)) ([]*gitlab.Note, error)
...
+	return events, nil
+}
+
+func readMergeRequestResourceStateEvents(it func() ([]*gitlab.ResourceStateEvent, error)) ([]*gitlab.ResourceStateEvent, error) {
+	var events []*gitlab.ResourceStateEvent
+
+	for {
+		page, err := it()
+		if err != nil {
+			return nil, errors.Wrap(err, "retrieving rse page")
```
Comment (eseliger) [inline cmd/repo-updater/repos/gitlab.go:567]:
```suggestion
			return nil, errors.Wrap(err, "retrieving resource state events page")
```

File: enterprise/internal/campaigns/webhooks_gitlab.go [hunk 5]
```diff
@@ -173,20 +174,64 @@ func (h *GitLabWebhook) handleEvent(ctx context.Context, extSvc *types.ExternalS
...
+		}
+		return nil
+
+	case *webhooks.MergeRequestUndraftEvent:
+		event := e.ToEvent()
+		pr := gitlabToPR(&e.Project, e.MergeRequest)
+		if err := h.upsertChangesetEvent(ctx, esID, pr, event); err != nil {
+			return &httpError{
+				code: http.StatusInternalServerError,
+				err:  errors.Wrap(err, "upserting changeset event"),
```
Comment (eseliger) [inline enterprise/internal/campaigns/webhooks_gitlab.go:234]: todo merge these

File: internal/extsvc/gitlab/notes.go [hunk 6]
```diff
@@ -53,10 +63,12 @@ func (c *Client) GetMergeRequestNotes(ctx context.Context, project *Project, iid
...
 const (
-	SystemNoteBodyReviewApproved               SystemNoteBody = "approved this merge request"
-	SystemNoteBodyReviewUnapproved             SystemNoteBody = "unapproved this merge request"
-	SystemNoteBodyReviewUnmarkedWorkInProgress SystemNoteBody = "unmarked as a **Work In Progress**"
-	SystemNoteBodyReviewMarkedWorkInProgress   SystemNoteBody = "marked as a **Work In Progress**"
+	SystemNoteBodyReviewApproved         SystemNoteBody = "approved this merge request"
+	SystemNoteBodyReviewUnapproved       SystemNoteBody = "unapproved this merge request"
+	SystemNoteBodyUnmarkedWorkInProgress SystemNoteBody = "unmarked as a **Work In Progress**"
+	SystemNoteBodyMarkedWorkInProgress   SystemNoteBody = "marked as a **Work In Progress**"
+	SystemNoteBodyMarkedDraft            SystemNoteBody = "marked this merge request as **draft**"
```
Comment (eseliger) [inline internal/extsvc/gitlab/notes.go:70]: gitlab.com changed these. :|

File: migrations/frontend/1528395764_drop_old_gitlab_events.up.sql [hunk 7]
```diff
@@ -0,0 +1,5 @@
+BEGIN;
+
+DELETE FROM changeset_events WHERE kind LIKE 'gitlab:%';
```
Comment (eseliger) [inline migrations/frontend/1528395764_drop_old_gitlab_events.up.sql:3]: The keys changed for the events, since we need to do truncation of the dates. I don't see a better way that doesn't cost a fortune. I think this is fine, though since we sync those MRs again anyways.

File: enterprise/internal/campaigns/webhooks_gitlab.go [hunk 8]
```diff
@@ -173,20 +168,26 @@ func (h *GitLabWebhook) handleEvent(ctx context.Context, extSvc *types.ExternalS
...
-	// All other merge request events are state events.
-	case stateMergeRequestEvent:
-		if err := h.handleMergeRequestStateEvent(ctx, esID, e); err != nil {
+	case *webhooks.MergeRequestCloseEvent,
+		*webhooks.MergeRequestMergeEvent,
+		*webhooks.MergeRequestReopenEvent,
+		*webhooks.MergeRequestDraftEvent,
+		*webhooks.MergeRequestUndraftEvent:
+		eventCommon := e.(webhooks.MergeRequestEventCommonContainer).ToEventCommon()
+		event := e.(webhooks.MergeRequestEventToEvent).ToEvent()
```
Comment (mrnugget) [inline enterprise/internal/campaigns/webhooks_gitlab.go:185]: Does it make sense to define an interface that combines these two so you can simplify the `case` here and don't need the casts?

File: enterprise/internal/campaigns/webhooks_gitlab_test.go [hunk 9]
```diff
@@ -649,35 +649,6 @@ func testGitLabWebhook(db *sql.DB, userID int32) func(*testing.T) {
 			})
 		})
 
-		t.Run("handleMergeRequestStateEvent changeset upsert error", func(t *testing.T) {
```
Comment (mrnugget) [inline enterprise/internal/campaigns/webhooks_gitlab_test.go:652]: What's the reason for removing that? It only checked whether an error bubbled up, right?

File: internal/extsvc/gitlab/merge_requests.go [hunk 10]
```diff
@@ -50,14 +50,17 @@ type MergeRequest struct {
 	// to do it the old fashioned way with lots of REST requests.
 	Notes     []*Note
 	Pipelines []*Pipeline
+	Events    []*ResourceStateEvent
```
Comment (mrnugget) [inline internal/extsvc/gitlab/merge_requests.go:53]: Nitpick/thought: should we rename `Events` to `ResourceStateEvents` so we don't have _our_ events and _their_ events further up and possibly get into naming confusing? I admit I am confused thinking about the taxonomy of GitLab events and our type definitions, but that has nothing to do with this PR, but maybe we can help that (me) by making the distinction clearer?

(Feel free to ignore though)

File: internal/repos/gitlab_test.go [hunk 11]
```diff
@@ -397,6 +401,8 @@ func TestGitLabSource_ChangesetSource(t *testing.T) {
 			p.changeset.Changeset.Metadata = mr
 			p.mockUpdateMergeRequest(mr, want, gitlab.UpdateMergeRequestStateEventReopen, nil)
 			p.mockGetMergeRequestNotes(mr.IID, nil, 20, nil)
+			// TODO: add event
```
Comment (mrnugget) [inline internal/repos/gitlab_test.go:404]: What does that mean?

File: migrations/frontend/1528395764_drop_old_gitlab_events.up.sql (hunk 7 above)
Comment (mrnugget) [reply migrations/frontend/1528395764_drop_old_gitlab_events.up.sql:3]: Any way to bump a timestamp on the gitlab changesets so that they're synced faster?

**Review by mrnugget — APPROVED** [review]

File: enterprise/internal/campaigns/webhooks_gitlab.go (hunk 8 above)
Comment (eseliger) [reply enterprise/internal/campaigns/webhooks_gitlab.go:185]: Might be, I wanted to be explicit though so we know which events we can upsert and which we can't. 🤔 WDYT?

File: enterprise/internal/campaigns/webhooks_gitlab_test.go (hunk 9 above)
Comment (eseliger) [reply enterprise/internal/campaigns/webhooks_gitlab_test.go:652]: The method doesn't exist anymore.

Comment (mrnugget) [reply enterprise/internal/campaigns/webhooks_gitlab_test.go:652]: Well, that makes sense then ;)

File: enterprise/internal/campaigns/webhooks_gitlab.go (hunk 8 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/webhooks_gitlab.go:185]: Not sure I understand or whether I've missed something in the first place, but what I meant was: can't you do something like this?

```go
type resourceStateEvent interface {
	webhooks.MergeRequestEventCommonContainer
	webhooks.MergeRequestEventToEvent
}
```

and then here, you can do:

```suggestion
		eventCommon := e.ToEventCommon()
		event := e.ToEvent()
```

Comment (eseliger) [reply enterprise/internal/campaigns/webhooks_gitlab.go:185]: Ah I misunderstood you then 🤦 That should work!

File: internal/extsvc/gitlab/notes.go [hunk 12]
```diff
@@ -40,10 +50,10 @@ func (c *Client) GetMergeRequestNotes(ctx context.Context, project *Project, iid
 			return nil, errors.Wrap(err, "requesting notes page")
 		}
 
-		// If there's another page, this will be a URL. If there's not, then
+		// If there's another page, this will be a page number. If there's not, then
 		// this will be an empty string, and we can detect that next iteration
 		// to short circuit.
-		url = header.Get("X-Next-Page")
+		currentPage = header.Get("X-Next-Page")
```
Comment (LawnGnome) [inline internal/extsvc/gitlab/notes.go:56]: Oof. Not my finest work.

File: internal/extsvc/gitlab/notes.go (hunk 6 above)
Comment (LawnGnome) [reply internal/extsvc/gitlab/notes.go:70]: I, for one, am shocked that having to parse their natural language text was not future proof.

File: internal/extsvc/gitlab/webhooks/merge_requests.go [hunk 13]
```diff
@@ -31,58 +28,159 @@ import (
...
+
+type keyer interface {
+	Key() string
+}
+
+// MergeRequestEventCommonContainer is a common interface for types that embed
+// MergeRequestEvent to provide a method that can return the embedded
+// MergeRequestEvent.
+type MergeRequestEventToEvent interface {
+	ToEvent() keyer
```
Comment (LawnGnome) [inline internal/extsvc/gitlab/webhooks/merge_requests.go:62]: It feels like this type has drifted from its original purpose: would a name like `AsKeyer()` and an interface name like `MergeRequestEventKeyer` or something be closer to the intent?

File: internal/extsvc/gitlab/webhooks/merge_requests.go [hunk 14]
```diff
@@ -134,6 +234,13 @@ func (mre *mergeRequestEvent) downcast() (interface{}, error) {
 		return &MergeRequestUnapprovedEvent{e}, nil
 
 	case "update":
+		if prev, curr := e.Changes.Title.Previous, e.Changes.Title.Current; prev != "" && curr != "" {
+			if gitlab.IsWIP(prev) && !gitlab.IsWIP(curr) {
+				return &MergeRequestUndraftEvent{e}, nil
+			} else if !gitlab.IsWIP(prev) && gitlab.IsWIP(curr) {
+				return &MergeRequestDraftEvent{e}, nil
+			}
```
Comment (LawnGnome) [inline internal/extsvc/gitlab/webhooks/merge_requests.go:242]: :scream:

**Review by LawnGnome — APPROVED** [review]
Nice work! I like the cleanups in here from the original, fairly hacky event handling.

I share a couple of Thorsten's questions, but otherwise, this LGTM. Is there anything changelog worthy in here?
